### ISSUE TITLE
List conflicts in MSB3277 reference-version-conflict message

### ISSUE THREAD

**Issue by rainersigwald** [dotnet/msbuild#608]
This is to keep track of this feature request from Connect: http://connect.microsoft.com/VisualStudio/feedback/details/2619450/unclear-warning-msb3277-found-conflicts-between-different-versions-of-the-same-dependent-assembly-that-could-not-be-resolved

> During compile I've got a warning: C:\Program Files (x86)\MSBuild\14.0\bin\Microsoft.Common.CurrentVersion.targets(1819,5): 
> warning MSB3277: Found conflicts between different versions of the same dependent assembly that could not be resolved.
>  These reference conflicts are listed in the build log when log verbosity is set to detailed. 
> 
> I believe that compiler knows the name of assembly that has a problem, and it can be included in the messages without re-compile in diagnostics mode.
> 
> Could you please change the message to include  assemblies name, as it does in diagnostics mode.

This message is emitted in [ResolveAssemblyReferences](https://github.com/Microsoft/msbuild/blob/8b488bd508807f415193fe38d5084cd590931cfb/src/XMakeTasks/AssemblyDependency/ResolveAssemblyReference.cs#L1069).

**Comment by rainersigwald** [comment]
We discussed this in triage this week. We're in favor of clear, actionable error messages, but since the information is just a rebuild with `/v:diag` away, we really don't want to slow down the normal/success case by carrying around extra information just to support this error. A proposed solution to this problem should not slow down the no-error case.

**Comment by cdmihai** [comment]
It seems like the information might be available, but just not logged: https://github.com/Microsoft/msbuild/blob/72c12de51702f7e214f1eba481f2f3f93b885616/src/XMakeTasks/AssemblyDependency/ResolveAssemblyReference.cs#L1039-L1042

**Comment by rainersigwald** [comment]
_From @davkean in #1712:_

This error message has been the bane of .NET developers for years. Having worked on assembly resolution for ~6 years in the BCL team, I've diagnosed so many of these warnings for customers, and there's three issues that I see with it:

1. Developers don't know what this message means.

2. Developers once they know what it means, don't know how to fix it.

3. The warning is not actionable without building again with diagnostic build (if you can figure out how to do that). Even then you need to search for the issue in the log, and find the relevant portions.

Here's a typical log:

```
 There was a conflict between "Microsoft.Extensions.DependencyModel, Version=1.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60" and "Microsoft.Extensions.DependencyModel, Version=1.0.1.0, Culture=neutral, PublicKeyToken=adb9793829ddae60".
            "Microsoft.Extensions.DependencyModel, Version=1.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60" was chosen because it was primary and "Microsoft.Extensions.DependencyModel, Version=1.0.1.0, Culture=neutral, PublicKeyToken=adb9793829ddae60" was not.
            References which depend on "Microsoft.Extensions.DependencyModel, Version=1.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60" [C:\Users\James\.nuget\packages\microsoft.extensions.dependencymodel\1.0.0\lib\netstandard1.6\Microsoft.Extensions.DependencyModel.dll].
                C:\Users\James\.nuget\packages\microsoft.extensions.dependencymodel\1.0.0\lib\netstandard1.6\Microsoft.Extensions.DependencyModel.dll
                  Project file item includes which caused reference "C:\Users\James\.nuget\packages\microsoft.extensions.dependencymodel\1.0.0\lib\netstandard1.6\Microsoft.Extensions.DependencyModel.dll".
                    C:\Users\James\.nuget\packages\microsoft.extensions.dependencymodel\1.0.0\lib\netstandard1.6\Microsoft.Extensions.DependencyModel.dll
            References which depend on "Microsoft.Extensions.DependencyModel, Version=1.0.1.0, Culture=neutral, PublicKeyToken=adb9793829ddae60" [].
                C:\Users\James\.nuget\packages\microsoft.testplatform.testhost\15.0.0-preview-20170210-02\lib\netstandard1.5\Microsoft.TestPlatform.CrossPlatEngine.dll
                  Project file item includes which caused reference "C:\Users\James\.nuget\packages\microsoft.testplatform.testhost\15.0.0-preview-20170210-02\lib\netstandard1.5\Microsoft.TestPlatform.CrossPlatEngine.dll".
                    C:\Users\James\.nuget\packages\microsoft.testplatform.testhost\15.0.0-preview-20170210-02\lib\netstandard1.5\Microsoft.TestPlatform.CrossPlatEngine.dll
                    C:\Users\James\.nuget\packages\microsoft.testplatform.testhost\15.0.0-preview-20170210-02\lib\netstandard1.5\testhost.dll
       C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(1956,5): warning MSB3277: Found conflicts between different versions of the same dependent assembly that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed.
```

Note the warning states:

```
warning MSB3277: Found conflicts between different versions of the same dependent assembly that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed.
```

But really the relevant portion is 1#:

```
 There was a conflict between "Microsoft.Extensions.DependencyModel, Version=1.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60" and "Microsoft.Extensions.DependencyModel, Version=1.0.1.0, Culture=neutral, PublicKeyToken=adb9793829ddae60".
```
and 2#

```
    References which depend on "Microsoft.Extensions.DependencyModel, Version=1.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60" [C:\Users\James\.nuget\packages\microsoft.extensions.dependencymodel\1.0.0\lib\netstandard1.6\Microsoft.Extensions.DependencyModel.dll].
                C:\Users\James\.nuget\packages\microsoft.extensions.dependencymodel\1.0.0\lib\netstandard1.6\Microsoft.Extensions.DependencyModel.dll
                  Project file item includes which caused reference "C:\Users\James\.nuget\packages\microsoft.extensions.dependencymodel\1.0.0\lib\netstandard1.6\Microsoft.Extensions.DependencyModel.dll".
                    C:\Users\James\.nuget\packages\microsoft.extensions.dependencymodel\1.0.0\lib\netstandard1.6\Microsoft.Extensions.DependencyModel.dll
            References which depend on "Microsoft.Extensions.DependencyModel, Version=1.0.1.0, Culture=neutral, PublicKeyToken=adb9793829ddae60" [].
                C:\Users\James\.nuget\packages\microsoft.testplatform.testhost\15.0.0-preview-20170210-02\lib\netstandard1.5\Microsoft.TestPlatform.CrossPlatEngine.dll
                  Project file item includes which caused reference "C:\Users\James\.nuget\packages\microsoft.testplatform.testhost\15.0.0-preview-20170210-02\lib\netstandard1.5\Microsoft.TestPlatform.CrossPlatEngine.dll".
                    C:\Users\James\.nuget\packages\microsoft.testplatform.testhost\15.0.0-preview-20170210-02\lib\netstandard1.5\Microsoft.TestPlatform.CrossPlatEngine.dll
                    C:\Users\James\.nuget\packages\microsoft.testplatform.testhost\15.0.0-preview-20170210-02\lib\netstandard1.5\testhost.dll
```

We should promote 1# or something like it as the root message, and 2# as a nested message under the warning. VS will now hide by default messages that too long or contain multiple lines, and let the user expand to see more detail.

**Comment by TehWardy** [comment]
Could we perhaps have an option to right click on the warning and select something like "Generate Binding Re-directions in Config" then if msbuild was to honor those redirections as "the will of the developer" then conflicts can be deemed to have been at least thought about and a solution provided.

**Comment by DarinMacRae** [comment]
@TehWardy Good suggestion. I think that is a really good first step.

**Comment by TehWardy** [comment]
@ewahner  It's basically a resolution in the build process / compiler where it finds multiple versions of a given assembly / doesn't find the thing it expects in some default location. 
The build process is technically correct in that the issue is how / what has been referred to is being "implied" in some way but the process has enough enough information to make an "educated guess" so it shows up as a warning rather than an error in the process.

The issue here of course is that resolving this problem properly is quite complex and actually requires a developer to make a decision but putting assembly redirection directives in the config file for the assembly you are building should resolve it.

In short:
Don't worry about it, it's just a frustration for those of us looking to keep our build reports clean!

**Comment by da3dsoul** [comment]
Just hit one of these for the first time. This issue helped me find it at least. +1 I guess

**Comment by dodegaard** [comment]
Thank you @davkean for your initial words on this frustrating issue. I have devs in VS Code who can't focus on actual warnings that matter due to this.  It used to be just the devops issue but clean coders obsess over the build log.  How can we lessen the noise in VS2017 or VS Code on this error? TIA.

**Comment by theunrepentantgeek** [comment]
From reading the source for `ResolveAssemblyReference.LogResults()`, I'd like to suggest the following incremental change.

Instead of generating a single warning that contains no specifics (the current behavior), generate a separate warning message for each distinct reference conflict.

For example:
```
MSB3277: Found conflicts between different versions of "Newtonsoft.Json" that could not be resolved.  
These reference conflicts are listed in the build log when log verbosity is set to detailed.
```

I've generally found that knowing the name of the conflicting assembly has been enough to track down the problem and address the issue.

All of the existing diagnostics would still be available in the detailed log if required, and I believe all of the information required for this change is already available within the `LogResults()` method, making this a simple change.

If this is an acceptable modification, I'm willing to have a go at making the change and creating a PR - but didn't want to dive in if there's another/better fix already underway.

**Comment by TehWardy** [comment]
@theunrepentantgeek that could be the key if the message included the project names that had the conflict too.

The key to solving this is knowing project + reference I believe.

**Comment by theunrepentantgeek** [comment]
Conflicts don't always stem from a direct project reference - sometimes there are intermediate dependencies on the path - this is why it's not always straightforward to solve.. That said, I'll have a look at what other information is on hand in `LogResults()`.

**Comment by TehWardy** [comment]
Yeh the trick is figuring out the root project in a tree I guess ... The ones that's the "app" is likely to be the one that needs something like binding redirection that would solve it for all of the projects compiled in to its build output directory.

Hence my suggestion above about making this process as painless to resolve as possible (with a simple right click).

As you say ... All the information is there right!

**Comment by kdawg1406** [comment]
Please fix this.  Opened over two years ago.

**Comment by jnm2** [comment]
Well- opened 1.5 years ago unless I'm missing something.

**Comment by cdmihai** [comment]
@theunrepentantgeek did #2379 which prints the name of the conflicting assembly. Let us know the extent to which this helps you diagnose and fix conflicts.

**Comment by kdawg1406** [comment]
@cdmihai looking forward to VS 15.5 RTM.

**Comment by Ofaim** [comment]
Well, as a noob, that fix is great for me!!
I was struggling the old warning with no success, but now I know where the main "conflict" is!
Furthermore, now the warning is refined - enabling us (developers) to discuss each "conflict" specifically (in many cases, the conflict emerged from the same situation/update action).
Thank you!!

**Comment by mfeingol** [comment]
On this topic... once you have a project with MSB3277 warnings... what are the next steps? Is there any formal guidance on this topic?

In my case, I have a UWP app with a number of NuGet package dependencies, and building it raises a number of these warnings. However, I'm not entirely sure what to do to address the problem, since my NuGet packages are generally up to date. Is there even a problem?

Thanks.

**Comment by Reminic** [comment]
One good reason to add more detailled infos in the message regarding the conflict: solving it may require setting the verbose log level to detailed, which ain't available to VS devs using VB.NET (src: https://docs.microsoft.com/en-us/dotnet/visual-basic/reference/command-line-compiler/verbose). Ouch !

**Comment by martinsuchan** [comment]
Using VS 15.7 and still seeing these warnings in one of our UWP solution. No idea, where or how to fix this:  
```
1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(2106,5): warning MSB3277: Found conflicts between different versions of "System.ComponentModel.EventBasedAsync" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed.
1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(2106,5): warning MSB3277: Found conflicts between different versions of "System.Diagnostics.Tools" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed.
1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(2106,5): warning MSB3277: Found conflicts between different versions of "System.Diagnostics.Debug" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed.
1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(2106,5): warning MSB3277: Found conflicts between different versions of "System.Threading" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed.
1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(2106,5): warning MSB3277: Found conflicts between different versions of "System.ComponentModel" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed.
```

**Comment by TehWardy** [comment]
@martinsuchan as stated above ... the fix is to add binding re-directions to your config files "manually" ... 
The normal turn around time for bugs in anything MS related from what I have seen is about 4 years so we have about another 2 ish still to wait on this.

**Comment by syndicatedshannon** [comment]
My gosh, this issue is so painful. Been solving it for years and it still pains me. Just fighting an instance of it where two assemblies have the same NuGet packages, the same references, the same project dependencies, the same output type, the same bindingRedirects... I can't freaking find the problem for hours and hours. SO painful. Grrr!!!

I page down through the diagnostic output and the scroll bar doesn't even move a pixel. I search the diagnostic output and there are literally hundreds of emissions of "System.Net.Http". And not one of them seems to correlate to "Found conflicts ... that could not be resolved.", because every occurrence lists which assembly was chosen and why.

Add my voice to those making an appeal to resolve this.

**Comment by syndicatedshannon** [comment]
A project with both <bindingRedirect oldVersion"4.0.0.0-4.2.0.0" newVersion="4.2.0.0"/> and project reference Version=4.2.0.0 chose Version=4.1.1.2 "because it was primary", due apparently to a hint path, causing report of MSB3277. Next, removing dependencies on obvious sources of Version=4.2.0.0 didn't fix it, because ApplicationInsights were enabled. Next, while collection/redirection up to 4.2.0.0 had no effect, collection/redirection down to 4.1.1.2 did. Lastly, simply adding a reference to a NetStandard project, despite having no dependence on the library in question, triggered override of the project primary version to the copy in Microsoft.NET.Build.Extensions via ImplicitlyExpandNETStandardFacades.

With so many counter-intuitive dependency calculations, this error really needs clarification to prevent wasting time.

**Comment by jmfallecker** [comment]
I'll also add comments here that this should be better presented. If the problem is known, why can't you offer an easier explanation of how to solve it?

The annoyance is that it's definitely solvable, but the steps to get there are unclear.

**Comment by TehWardy** [comment]
Some light guys ...
I've noticed this happening less and the one suggesting adding a binding redirect happening more in my solutions (it might be that stuff is going on behind the scenes to make that happen).

I have also found that managing nuget packages for the solution and consolidating references often helps.
Not a complete solution and may still require some manual removal / updates of binding redirections but it's a start.

IMO this should be trivial to fix but I guess there's more to it than we see going on behind the scenes.
My worst nightmare being when it happens on a build server but not locally.

**Comment by cdmihai** [comment]
@TehWardy Check out [CentralPackageVersions](https://github.com/Microsoft/MSBuildSdks/tree/master/src/CentralPackageVersions) to help you specify all your versions in one place.

**Comment by dsplaisted** [comment]
We should consider adding the suggested redirects to this message if we're not going to do that, then this issue can be closed (as we now list the assembly name with the conflict).

**Comment by TehWardy** [comment]
@cdmihai doesn't work in my case ...
Do anything with say ... Azure's fluent API and some assemblies force older versions of packages.

I'd like to for something like say Newtonsoft.Json just put a rule in to force latest version across my entire codebase, but something like Azure's Functions SDK requires v9 for some weird reason.

It's stuff like this that triggers most my frustration with MS as an API provider when they are also responsible for the tools consuming those assemblies problems like this remind me of the "dll hell" that Java / game devs often complain about.

**Comment by dcb99** [comment]
This just popped up in a project I'm working on. I added an xUnit test project and when I try and build I get ``C:\Program Files\dotnet\sdk\2.2.105\Microsoft.Common.CurrentVersion.targets(2110,5): warning MSB3277: Found conflicts between different versions of "Microsoft.AspNetCore.Diagnostics.Abstractions" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. [C:\Users\dbarless\Desktop\MaintReq\MaintReq.tests\MaintReq.tests.csproj]
C:\Program Files\dotnet\sdk\2.2.105\Microsoft.Common.CurrentVersion.targets(2110,5): warning MSB3277: Found conflicts between different versions of "Microsoft.AspNetCore.Mvc" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. [C:\Users\dbarless\Desktop\MaintReq\MaintReq.tests\MaintReq.tests.csproj]
C:\Program Files\dotnet\sdk\2.2.105\Microsoft.Common.CurrentVersion.targets(2110,5): warning MSB3277: Found conflicts between different versions of "Microsoft.AspNetCore.Mvc.Formatters.Json" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. [C:\Users\dbarless\Desktop\MaintReq\MaintReq.tests\MaintReq.tests.csproj]
C:\Program Files\dotnet\sdk\2.2.105\Microsoft.Common.CurrentVersion.targets(2110,5): warning MSB3277: Found conflicts between different versions of "Microsoft.AspNetCore.Mvc.ViewFeatures" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. [C:\Users\dbarless\Desktop\MaintReq\MaintReq.tests\MaintReq.tests.csproj]
C:\Program Files\dotnet\sdk\2.2.105\Microsoft.Common.CurrentVersion.targets(2110,5): warning MSB3277: Found conflicts between different versions of "Microsoft.AspNetCore.Razor.Runtime" that could not be resolved.  These reference conflicts are listed in the build log
when log verbosity is set to detailed. [C:\Users\dbarless\Desktop\MaintReq\MaintReq.tests\MaintReq.tests.csproj]

What?!? How does this get resolved, why is it happening? I've never had this happen with any other project. It's annoying that this is apparently a problem that's existed for 2 years with no clear resolution.

**Comment by syndicatedshannon** [comment]
IMO logging the assembly in conflict is helpful, but not sufficient, since tracing the conflict with only that information can be very time consuming, or lead the developer outside their own code domain, in my past experience even terminating in workarounds for defects in MS-provided tools and assemblies.

Ideally, a deterministic strategy to resolve this quickly would exist, and the build process would emit information in support of that process. For example, perhaps the log would trace the origin of the conflict; although I recognize that MSBuild limits what it holds for performance reasons.

**Comment by StingyJack** [comment]
I for one would love to stop getting these kinds of warning walls.   Do note that this is just the top half of the warning.

![image](https://user-images.githubusercontent.com/15381181/68805751-6e7c6900-0632-11ea-871c-97ea77efea85.png)

**Comment by TehWardy** [comment]
@StingyJack  that' the good version of this problem ... if I recall you should be able to just double click on that and it'll dump that XML blob in to your config file.

**Comment by StingyJack** [comment]
@TehWardy except that maybe a years of users testing code built from projects that exhibit the wall of warning leads me to believe that they are false positives.

**Comment by TehWardy** [comment]
The issue is what the build process sees vs what actually gets deployed ... 
Testers would need to be testing lines of code in your codebase that specifically calls the edge case updates in the dependencies that are in question for there to actually be a problem.

I would think that the reason your testers aren't finding anything is that you either don't use those edge cases in the dependent libs or they have an incomplete test case suite.

**Comment by StingyJack** [comment]
This happens when reffing a netstandard2.0 assembly from a 4.6.2 project.

**Comment by jnm2** [comment]
@StingyJack Windows 7 is gone in January. Assuming everyone is on Windows 10 at that point, which forces you to stay up to date within at most a year, everyone has 4.7.2 already and soon will 4.8 will be the new baseline.

**Comment by StingyJack** [comment]
> We're in favor of clear, actionable error messages, but since the information **is just a rebuild with `/v:diag` away**, we really don't want to slow down the normal/success case by carrying around extra information just to support this error.

@rainersigwald   - you mean its just an opening of Tools -> Options, then Projects and Solutions -> Build and Run, then select the Diagnostic option for the two verbosity options, click OK, then rebuild , and then sift through **10's of thousands** of lines of log output (if not more) to find the problem source?  

Thats a total PITA compared to the message just telling us what the problem is. Why make your customers work so much when its possible for you to improve their experience with this cryptic and questionable warning?

At least this used to explains how to find the source. I see this has been changed at some point and now we dont get that instruction, and instead get a message like...

```
Warning		Found conflicts between different versions of the same dependent assembly. 
Please set the "AutoGenerateBindingRedirects" property to true in the project file. 
For more information, see http://go.microsoft.com/fwlink/?LinkId=294190.	
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets	
2081	
```

... for projects that have that flag set to True in the project file.
